harness/tests/launch/test_autodeepspeed.py

```diff
@@ -0,0 +1,359 @@
+
+    assert mock_check_sshd.call_count == len(cluster_info.container_addrs)
+    mock_check_sshd.assert_has_calls(
+        [
+            mock.call(addr, mock_start_time + 20, constants.DTRAIN_SSH_PORT)
+            for addr in cluster_info.container_addrs
+        ]
+    )
+
+    launch_proc_mock().wait.assert_called_once()
+
+    sshd_proc_mock().kill.assert_called_once()
+    sshd_proc_mock().wait.assert_called_once()
+
+
+@mock.patch("subprocess.Popen")
+@mock.patch("determined.get_cluster_info")
+@mock.patch("determined.util.check_sshd")
+@mock.patch("time.time")
+def test_launch_multi_slot_fail(
+    mock_time: mock.MagicMock,
+    mock_check_sshd: mock.MagicMock,
+    mock_cluster_info: mock.MagicMock,
+    mock_subprocess: mock.MagicMock,
+) -> None:
+    cluster_info = make_mock_cluster_info(["0.0.0.0", "0.0.0.1"], 0)
+    mock_cluster_info.return_value = cluster_info
+    mock_start_time = time.time()
+    mock_time.return_value = mock_start_time
+    mock_check_sshd.side_effect = ValueError("no sshd greeting")
+
+    train_entrypoint = "model_def:TrialClass"
+    sshd_cmd = [
+        "/usr/sbin/sshd",
+        "-p",
+        str(constants.DTRAIN_SSH_PORT),
+        "-f",
+        "/run/determined/ssh/sshd_config",
+        "-D",
+    ]
+
+    pid_server_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.pid_server",
+        "--on-fail",
+        "SIGTERM",
+        "--on-exit",
+        "SIGTERM",
+        "--grace-period",
+        "5",
+        f"/tmp/pid_server-{cluster_info.allocation_id}",
+        str(len(cluster_info.slot_ids)),
+        "--",
+    ]
+
+    deepspeed_cmd = [
+        "deepspeed",
+        "-H",
+        autodeepspeed.hostfile_path,
+        "--master_addr",
+        cluster_info.container_addrs[0],
+        "--no_python",
+        "--no_local_rank",
+        "--",
+    ]
+
+    pid_client_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.pid_client",
+        f"/tmp/pid_server-{cluster_info.allocation_id}",
+        "--",
+    ]
+
+    log_redirect_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.worker_process_wrapper",
+        "RANK",
+        "--",
+    ]
+
+    harness_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.harness",
+        "--train-entrypoint",
+        train_entrypoint,
+    ]
+
+    launch_cmd = pid_server_cmd + deepspeed_cmd + pid_client_cmd + log_redirect_cmd + harness_cmd
+
+    sshd_proc_mock = mock.MagicMock()
+    launch_proc_mock = mock.MagicMock()
+
+    def mock_process(cmd: List[str], *args: Any, **kwargs: Any) -> Any:
+        if cmd == sshd_cmd:
+            return sshd_proc_mock(*args, **kwargs)
+        if cmd == launch_cmd:
+            return launch_proc_mock(*args, **kwargs)
+        return None
+
+    mock_subprocess.side_effect = mock_process
+
+    with set_container_id_env_var():
+        with pytest.raises(ValueError):
```

Comment:
can you include a `matches=` arg here, so we don't get a false positive?  Also it would make the test easier to understand when reading it, so it's a double-win